### PR TITLE
Re-organize C++ API `torch::nn` folder structure

### PR DESCRIPTION
**Pull request by yf225**
This PR aims to re-organize C++ API `torch::nn` folder structure in the following way:
- Every module in `torch/csrc/api/include/torch/nn/modules/` (except `any.h`, `named_any.h`, `modulelist.h`, `sequential.h`, `embedding.h`) has a strictly equivalent Python file in `torch/nn/modules/`. For  example:
`torch/csrc/api/include/torch/nn/modules/pooling.h` -> `torch/nn/modules/pooling.py`
`torch/csrc/api/include/torch/nn/modules/conv.h` -> `torch/nn/modules/conv.py`
`torch/csrc/api/include/torch/nn/modules/batchnorm.h` -> `torch/nn/modules/batchnorm.py`
`torch/csrc/api/include/torch/nn/modules/sparse.h` -> `torch/nn/modules/sparse.py`
- Containers such as  `any.h`, `named_any.h`, `modulelist.h`, `sequential.h` are moved into `torch/csrc/api/include/torch/nn/modules/container/`, because their implementations are too long to be combined into one file (like `torch/nn/modules/container.py` in Python API)
- `embedding.h` is not renamed to `sparse.h` yet, because we have another work stream that works on API parity for Embedding and EmbeddingBag, and renaming the file would cause conflict. After the embedding API parity work is done, we will rename `embedding.h` to  `sparse.h` to match the Python file name, and move the embedding options out to options/ folder.
- `torch/csrc/api/include/torch/nn/functional/` is added, and the folder structure mirrors that of `torch/csrc/api/include/torch/nn/modules/`. For example, `torch/csrc/api/include/torch/nn/functional/pooling.h` contains the functions for pooling, which are then used by the pooling modules in `torch/csrc/api/include/torch/nn/modules/pooling.h`.
- `torch/csrc/api/include/torch/nn/options/` is added, and the folder structure mirrors that of `torch/csrc/api/include/torch/nn/modules/`. For example, `torch/csrc/api/include/torch/nn/options/pooling.h` contains MaxPoolOptions, which is used by both MaxPool modules in `torch/csrc/api/include/torch/nn/modules/pooling.h`, and max_pool functions in `torch/csrc/api/include/torch/nn/functional/pooling.h`.

### CODE REVIEW

**Comment by yf225** [comment]
@ShahriarSS Would you like to review this PR?

File: torch/csrc/api/include/torch/nn/functional/pooling.h [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <torch/nn/options/pooling.h>
+
+namespace torch {
+namespace nn{
+namespace functional {
+
+inline Tensor avg_pool1d(const Tensor& input, AvgPool1dOptions options) {
```
Comment (ShahriarSS) [inline torch/csrc/api/include/torch/nn/functional/pooling.h]: @yf225 Shouldn't options be ```const &``` in all of the functions in this file?

**Review by ShahriarRezghi — COMMENTED** [review]
@yf225 Also are you planning on separating options from ```conv``` and ```fold``` and others too like you did with avgpool and maxpool?

**Review by ShahriarRezghi — COMMENTED** [review]
@yf225 in ``` torch/csrc/api/include/torch/nn/modules/linear.cpp``` line 11 you have left ```LinearOptions``` contructor in its place instead of moving it to ```torch/csrc/api/include/torch/nn/options/linear.h```. Is this intentional?
Also should we add ```fold``` function to ```torch::nn::functional```?

**Comment by yf225** [comment]
@ShahriarSS Thanks a lot for the catch! I moved the other options as well. I will add `fold` function and other functions to `torch::nn::functional` in the next PR.

**Comment by ShahriarSS** [comment]
@yf225 This looks good to me.

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**